compiler/src/test/java/com/squareup/anvil/compiler/internal/reference/ClassReferenceTest.kt

```diff
@@ -76,6 +77,85 @@ class ClassReferenceTest {
       assertThat(exitCode).isEqualTo(OK)
     }
   }
+
+  @Test fun `the type parameter list is parsed`() {
+    compile(
+      """
+      package com.squareup.test
+
+      class SomeClass1<T : List<String>>(
+        private val t: T
+      )
+      
+      class SomeClass2<V>(
+        private val v: V
+      ) where V : Appendable, V : CharSequence
+
+      class SomeClass3<T, R : Set<String>>(
+        private val t: T,
+        private val r: Lazy<R>
+      ) where T : Appendable, T : CharSequence
+      """,
+      allWarningsAsErrors = false,
+      codeGenerators = listOf(
+        simpleCodeGenerator { psiRef ->
+          val descriptorRef = psiRef.toDescriptorReference()
+
+          when (psiRef.shortName) {
+            "SomeClass1" -> {
+              val psiParameter = psiRef.typeParameters.single()
+              val descriptorParameter = descriptorRef.typeParameters.single()
+
+              assertThat(psiParameter.name).isEqualTo("T")
+              assertThat(descriptorParameter.name).isEqualTo("T")
+
+              assertThat(psiParameter.upperBounds.single().asTypeName().toString())
+                .isEqualTo("kotlin.collections.List<kotlin.String>")
+              assertThat(descriptorParameter.upperBounds.single().asTypeName().toString())
+                .isEqualTo("kotlin.collections.List<kotlin.String>")
+            }
+            "SomeClass2" -> {
+              val psiParameter = psiRef.typeParameters.single()
+              val descriptorParameter = descriptorRef.typeParameters.single()
+
+              assertThat(psiParameter.name).isEqualTo("V")
+              assertThat(descriptorParameter.name).isEqualTo("V")
+
+              assertThat(psiParameter.upperBounds[0].asClassReference().fqName)
+                .isEqualTo(FqName("java.lang.Appendable"))
+              assertThat(psiParameter.upperBounds[1].asClassReference().fqName)
+                .isEqualTo(FqName("kotlin.CharSequence"))
+
+              assertThat(descriptorParameter.upperBounds[0].asClassReference().fqName)
+                .isEqualTo(FqName("java.lang.Appendable"))
+              assertThat(descriptorParameter.upperBounds[1].asClassReference().fqName)
+                .isEqualTo(FqName("kotlin.CharSequence"))
+            }
+            "SomeClass3" -> {
+              listOf(psiRef, descriptorRef).forEach { ref ->
```

Comment:
I like how you used the list to avoid duplicating the asserts between each ref here. Any reason you decided not to do the same for the other test cases?

Reply:
Laziness 🙃  I should have done this. In general I don't like loops, because if the test fails then you don't know what the arguments are for the current iteration. But I got tired of duplicating all code, so I started using the loop.